doc/adr/0018-synchronising-queries-across-indexers-running-in-parallel.rst

```diff
@@ -0,0 +1,65 @@
+
+.. synchronising_queries_across_indexers_running_in_parallel:
+
+ADR 18: Synchronising queries across indexers running in parallel
+=================================================================
+
+Date: 2023-02-24
+
+Authors
+-------
+
+Ometita Radu - Adrian <radu.ometita@iohk.io>
+
+Status
+------
+
+Draft
+
+Context
+-------
+
+Running two indexers in parallel creates the implicit assumption that both indexers are answering queries over the same original (blockchain) data. That is not generally true due to the different speeds at which indexers process blocks.
+
+This has been observed first when the PAB needed to index some blockchain information and a smart contract would query both the PAB and the Chain Index. It would often be the case that their views were not synchronised, requiring the developer to synchronise the two views, which is not a trivial thing to do.
+
+Currently, the way we almost avoid that situation is to run a set of indexers in parallel by waiting on all of them to finish processing the current block before advancing to the next one. This is a fairly big improvement, but essentially it still has the same problem, except that it will not be triggered as often. One scenario when this could happen is if the query comes while the indexers are about to process a block. Some of these indexers may answer the query before block processing, and some after the block has been processed. This can be mitigated by a query-locking mechanism that spans all the indexers, but this adds complexity while not managing to solve the root problem.
+
+This solution also forgoes the possibility of splitting up the indexers into different processes or machines. Our customers complained about the excessive resource consumption, performance and the inability to easily add their own indexers to the chain index. Giving users the possibility of distributing their indexers will partially address the first concern by splitting resource consumption on several machines (which is also addressed at the indexer design level by an inherently scalable design) and indirectly the second (since indexers are distributable, they have a simple, standard API that allows any customer to roll their own version, implicitly solving the hard problems like rollbacks and synchronisation).
+
+Decision
+--------
+
+1. Change the `Request` type to include a preferred slot number and how many results we want to receive. If there is no preferred slot number, the top available slot number is assumed.
```

Comment:
Why include in the request the number of results we want to receive?

Reply:
I am assuming that different applications will have different synchronisation requirements, so putting that number in the query seems like a good way to provide that.

Reply:
Ah so you're not talking about a result limit (like the `LIMIT` SQL keyword)? Maybe paraphrase?